### PR TITLE
Add guidance about unknown fields

### PR DESCRIPTION
**Pull request by asraa**
https://github.com/theupdateframework/specification/issues/163

Addressing this, specify that implementers should preserve unknown key-value pairs and include them in calculating hashes or verifying signatures.

I would like to add examples here, but it feels like an awkward place if the format hasn't been defined at this point in the spec.

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by mnm678** [comment]
Thanks @asraa! The text of the change looks great, but could you remove the whitespace changes? They make the diff hard to read.

**Comment by asraa** [comment]
Sorry about that! Must've been auto-formatting.

File: tuf-spec.md [hunk 1]
```diff
@@ -490,9 +490,11 @@ A delegated role file is located at:
 # Document formats # {#document-formats}
 
 All of the formats described below include the ability to add more
-attribute-value fields for backwards-compatible format changes.  If
-a backwards incompatible format change is needed, a new filename can
-be used.
+attribute-value fields to dictionary objects for backwards-compatible format
```
Comment (trishankatdatadog) [inline tuf-spec.md:493]:
```suggestion
key-value fields to dictionary objects for backwards-compatible format
```

File: tuf-spec.md [hunk 2]
```diff
@@ -490,9 +490,11 @@ A delegated role file is located at:
...
+changes.  Implementers who encounter undefined key-value pairs in the format
+must include the data when calculating hashes or verifying signatures and must
+preserve the data when re-serializing. If a backwards incompatible format change
```
Comment (trishankatdatadog) [inline tuf-spec.md:496]: I'm not sure what this last sentence means exactly, but we'll leave it for another PR, since you didn't write it. Maybe someone should make an issue to try to clarify it, though.

**Review by trishankatdatadog — COMMENTED** [review]
I can't think of how this PR affects security, at least for TUF purposes. Did I miss anything? @mnm678 @SantiagoTorres @JustinCappos

File: tuf-spec.md (hunk 2 above)
Comment (mnm678) [reply tuf-spec.md:496]: This looks like a sentence left from pre-semantic versioning. I opened #170.

**Comment by mnm678** [comment]
> I can't think of how this PR affects security, at least for TUF purposes. Did I miss anything?

Well, someone could try some kind of code injection through the unknown field, but it's signed metadata, so that's pretty unlikely.

**Comment by asraa** [comment]
> Well, someone could try some kind of code injection through the unknown field, but it's signed metadata, so that's pretty unlikely.

This would be like someone adding an strange field that disappears when canonicalizing and verifying the signature, right? Did this ever have some discussion with regards to the CUSTOM targets dictionary? I suppose one mitigation is that users must consume canonicalized custom key-values or something like that.

**Comment by trishankatdatadog** [comment]
> This would be like someone adding an strange field that disappears when canonicalizing and verifying the signature, right? Did this ever have some discussion with regards to the CUSTOM targets dictionary? I suppose one mitigation is that users must consume canonicalized custom key-values or something like that.

I think Marina means something else, but I'm not sure what

**Comment by mnm678** [comment]
I was thinking of a json parsing vulnerability. But any of these attacks would require an attacker to tamper with signed metadata, so there shouldn't be a security implication.

**Comment by trishankatdatadog** [comment]
> I was thinking of a json parsing vulnerability. But any of these attacks would require an attacker to tamper with signed metadata, so there shouldn't be a security implication.

Right, and any attacker who controls the key can do this without unknown key-value fields anyway.

Asra, what attack did you have in mind?

**Comment by asraa** [comment]
> Asra, what attack did you have in mind?

I thought similar -- as in an attacker adds some specially crafted fields that when canonicalized disappear, so the signature is still verified, by a client thinks there's some extra "verified" data. This may not be possible though (and is obviously visible in the files)

**Comment by trishankatdatadog** [comment]
> I thought similar -- as in an attacker adds some specially crafted fields that when canonicalized disappear, so the signature is still verified, by a client thinks there's some extra "verified" data. This may not be possible though (and is obviously visible in the files)

Yeap, the good news is that they will be caught, provided we log everything that goes in and out 🙂 

The safest thing to do is to explicitly disallow superset fields, but I'm afraid this is too restrictive for valid, real-world use cases.

I can see how TUF metadata can be used to carry other things that might lead to insecurity, but I don't see how that is directly a problem, so long as it does not break TUF's security guarantees. This is already doable with the targets themselves anyway, not to mention the custom targets metadata.

**Comment by trishankatdatadog** [comment]
BTW, this has implications for the [sigstore key ceremony](https://github.com/sigstore/root-signing) this Friday, which uses superset fields to record YubiKey attestations alongside public keys in their TUF root metadata. So, if anyone is really concerned, please speak up ASAP.

File: tuf-spec.md (hunk 1 above)
Comment (joshuagl) [reply tuf-spec.md]: The rest of the spec is fairly consistent in referring to them as attributes and attribute-value fields.

Comment (joshuagl) [reply tuf-spec.md]: I think for this PR we should revert to attribute-value fields, to be consistent.

We might want to edit the entire spec to change the object related terminology. Filed as https://github.com/theupdateframework/specification/issues/175

**Comment by trishankatdatadog** [comment]
I can't think of major security issues right now with this change. Does anyone see issues? @JustinCappos?

**Comment by joshuagl** [comment]
I can't think of a security issue with the change either, I did suggest we change some phrasing to be more consistent with the rest of the spec https://github.com/theupdateframework/specification/pull/169#discussion_r672284490 (but I also prefer the phrasing used in this PR, and think we should convert the entire spec to use it https://github.com/theupdateframework/specification/issues/175)

**Review by trishankatdatadog — DISMISSED** [review]
Approved, but I want everyone here to think about this

File: tuf-spec.md [hunk 3]
```diff
@@ -490,9 +490,11 @@ A delegated role file is located at:
 # Document formats # {#document-formats}
 
 All of the formats described below include the ability to add more
-attribute-value fields for backwards-compatible format changes.  If
-a backwards incompatible format change is needed, a new filename can
-be used.
+key-value fields to dictionary objects for backwards-compatible format
+changes.  Implementers who encounter undefined key-value pairs in the format
+must include the data when calculating hashes or verifying signatures and must
+preserve the data when re-serializing. If a backwards incompatible format change
```
Comment (joshuagl) [inline tuf-spec.md:496]:
```suggestion
attribute-value fields to objects for backwards-compatible format changes.
Implementers who encounter undefined attribute-value pairs in the format
must include the data when calculating hashes or verifying signatures and must
preserve the data when re-serializing. If a backwards incompatible format change
```

The spec consistently uses attribute-value and object (rather than key-value and dictionary). Let's use the standard terms until https://github.com/theupdateframework/specification/issues/175 is addressed.

**Review by joshuagl — APPROVED** [review]
LGTM, thanks @asraa – just a minor suggestion to use the specification standard terms.

**Review by mnm678 — APPROVED** [review]

**Review by trishankatdatadog — APPROVED** [review]

**Review by trishankatdatadog — APPROVED** [review]

**Review by mnm678 — APPROVED** [review]
